### PR TITLE
s3_url is and cast procs & s3_tools updates

### PR DESCRIPTION
**Pull request by adam-bethell**
## Changes
S3 subcommands have been updated to optionally use s3_url in place of bucket and remote_file (object_key)

### New Procs
qc::s3_url_bucket_object_key

qc::is s3_bucket
qc::is s3_object_key
qc::is s3_url

qc::cast s3_url
qc::castable s3_url

## Tests
Testing of s3_tools procs in progress
is.test updated with is proc tests

### CODE REVIEW

File: tcl/cast.tcl [hunk 1]
```diff
@@ -830,6 +830,30 @@ namespace eval qc::cast {
...
+        
+        # Force the url to start "S3://"
+        set s3_url $string
+        if {[regexp {^[sS]3://(.*)$} $s3_url -> temp]} {
+            set s3_url $temp
+        } elseif {[regexp {^/(.*)$} $s3_url -> temp]} {
+            set s3_url $temp
+        }
+        
+        return "S3://${$bucket}/{$object_key}"
```
Comment (bvw) [inline tcl/cast.tcl:854]: I think protocol is lower case by convention eg http, https, ftp , sftp etc

File: tcl/s3_tools.tcl [hunk 2]
```diff
@@ -245,9 +245,18 @@ proc qc::s3 { args } {
             if { [llength $args] < 3 || [llength $args] > 4 } {
                 error "Wrong number of arguments. Usage: qc::s3 get mybucket remote_filename {local_filename}"
             } elseif { [llength $args] == 3 } {
-                # No local filename, assume same as remote_filename in current directory
-                lassign $args -> bucket remote_filename 
-                set local_filename "./[file tail $remote_filename]"
+                # Test if args are $bucket and $remote_filename or $s3_url and $local_filename
+                lassign $args -> temp0 temp1
```
Comment (bvw) [inline tcl/s3_tools.tcl:249]: I suggest `arg0` `arg1` as variable names instead of `temp0` `temp1`

**Comment by adam-bethell** [comment]
## Updates
- Tests for S3 tools (and debugging of the procs themselves from the test results)
- Tests for `qc::cast s3_url`
- Fixed tests for `qc::is` `s3_url`, and `s3_object_key`

## TODO
- Test for the `qc::s3 upload <subcomman>` procs
- Tests for `qc::castable s3_url`
- Sort the naming conflict of `s3_url` to mean something like `s3://bucket/object_key` and s3_url to mean `bucket.s3.amazonaws.com`
  - The different is one is an URL with the S3 protocol. The other is an HTTP URL _for_ S3
  - One of the main aims in this PR is to not make any breaking changes therefore the new usage of s3_url needs to be renamed. But to what?

**Comment by adam-bethell** [comment]
## TODO
- [ ] Rename `is`, `cast`, `castable` procs for `s3_url` to `s3_uri`
- [ ] __Breaking Change:__ Remove option for passing in an a path and enforce object_key for non-subcommand s3 procs
  - Example: Update `qc::s3_get` to no longer take `/my-file` as an argument and instead `my-file` or an s3_uri where appropriate 
  - N.B: the command procs (e.g. `qc::s3 get`) will still accept paths 
- [ ] Add new s3 subcommand for constructing an s3 endpoint `qc::s3 endpoint` and replace instances of `qc::s3_url`
  - Example: `qc::s3 endpoint $bucket $object_key` => `bucket.s3.amazonaws.com/object_key`
  - Example: `qc::s3 endpoint $s3_uri` => `bucket.s3.amazonaws.com/object_key`
- [ ] Add new s3 subcommand for constructing an s3 URI `qc::s3 uri`
  - Example: `qc::s3 uri $bucket $object_key` => `s3://bucket/object_key`
  - Example: `qc::s3 uri $bucket`  => `s3://bucket`
  - N.B. to get an s3 URI from a string like "bucket/object_key" the proc `qc::cast s3_uri` would be used

**Comment by adam-bethell** [comment]
Updated

**Pull request merged**